framework/include/executioners/Runner.h

```diff
@@ -0,0 +1,165 @@
+      return s;
+    }
+
+    /// Marks the result as passing/converged with the given msg text
+    /// describing detail about how things ran.  A result object is in this
+    /// state by default, so it is not necessary to call this function for
+    /// converged/passing scenarios.
+    void pass(const std::string & msg, bool overwrite = false)
+    {
+      mooseAssert(converged || overwrite,
+                  "cannot override nonconverged executioner result with a passing one");
+      ((void)(overwrite)); // avoid unused error due to assert
+      reason = msg;
+      converged = true;
+    }
+
+    /// Marks the result as failing/unconverged with the given msg text
+    /// describing detail about how things ran.
+    void fail(const std::string & msg)
+    {
+      reason = msg;
+      converged = false;
+    }
+
+    /// Records results from sub/internal runners in a runner's result.  When
+    /// child-runners return a result object following their execution, this
+    /// function should be called to add that info into the result heirarchy.
+    /// If the child runner was identified by a label/text from the input file
+    /// (e.g. via sub_solve1=foo_runner) - then "name" should be "sub_solve1".
+    bool record(const std::string & name, const Result & r)
+    {
+      subs[name] = r;
+      return r.convergedAll();
+    }
+
+    /// Returns false if any single runner in the current hierarchy of results
+    /// (i.e. including all child results accumulated recursively via record)
+    /// had a faild/unconverged return state.  Returns true otherwise.  This
+    /// is how convergence should generally be checked/tracked by runners -
+    /// rather than accessing e.g. the converged member directly.
+    bool convergedAll() const
+    {
+      if (!converged)
+        return false;
+      for (auto & entry : subs)
+        if (!entry.second.convergedAll())
+          return false;
+      return true;
+    }
+
+  private:
+    std::string label(bool success_msg, const std::string & subname = "")
+    {
+      std::string state_str =
+          success_msg || !converged ? (std::string("(") + (converged ? "pass" : "FAIL") + ")") : "";
+      return subname + (subname.empty() ? "" : ":") + _name + state_str +
+             ((success_msg || !converged) && !reason.empty() ? ": " + reason : "");
+    }
+    std::string _name;
+  };
+
+  Runner(const InputParameters & parameters);
+
+  virtual ~Runner() {}
+
+  static InputParameters validParams();
+
+  /// This is the main function for runners - this is how runners should
+  /// invoke child/sub runners - by calling their run function.
+  Result run();
+
+  /// This function contains the primary execution implementation for a
+  /// runner.  Custom runner behavior should be localized to this function.  If
+  /// you are writing a runner - this is basically where you should put all your
+  /// code.
+  virtual Result gogogadget() = 0;
```

Comment:
Can we get a different name here?  Even for experimental code... it would be better to have something else.

Reply:
`runImplementation`, `actuallyRun`, `sprint`, `start`, `begin`, `doRun`, `executeRun`

Reply:
Or maybe just `run` and `[run]Inner`, or `flexecute` and `flexecuteInner`.  Regardless, I was planning on renaming this function as well - but I figured its name will be informed by the rename for the class/system anyway.

Reply:
I like `executeRun`